Source/Core/DolphinQt/FIFO/FIFOAnalyzer.cpp

```diff
@@ -139,189 +144,277 @@ void FIFOAnalyzer::UpdateTree()
+      "GX_DRAW_LINE_STRIP",   "GX_DRAW_POINTS",
+  };
+  return fmt::format("{} VAT {}", names[primitive], vat);
+}
+
 void FIFOAnalyzer::UpdateDetails()
 {
-  m_detail_list->clear();
+  // Clearing the detail list can update the selection, which causes UpdateDescription to be called
+  // immediately.  However, the object data offsets have not been recalculated yet, which can cause
+  // the wrong data to be used, potentially leading to out of bounds data or other bad things.
+  // Clear m_object_data_offsets first, so that UpdateDescription exits immediately.
   m_object_data_offsets.clear();
+  m_detail_list->clear();
+  m_search_results.clear();
+  m_search_next->setEnabled(false);
+  m_search_previous->setEnabled(false);
+  m_search_label->clear();
+
+  if (!FifoPlayer::GetInstance().IsPlaying())
+    return;
 
-  auto items = m_tree_widget->selectedItems();
+  const auto items = m_tree_widget->selectedItems();
 
   if (items.isEmpty() || items[0]->data(0, OBJECT_ROLE).isNull())
     return;
 
-  int frame_nr = items[0]->data(0, FRAME_ROLE).toInt();
-  int object_nr = items[0]->data(0, OBJECT_ROLE).toInt();
+  const u32 frame_nr = items[0]->data(0, FRAME_ROLE).toUInt();
+  const u32 object_nr = items[0]->data(0, OBJECT_ROLE).toUInt();
 
   const auto& frame_info = FifoPlayer::GetInstance().GetAnalyzedFrameInfo(frame_nr);
   const auto& fifo_frame = FifoPlayer::GetInstance().GetFile()->GetFrame(frame_nr);
 
-  const u8* objectdata_start = &fifo_frame.fifoData[frame_info.objectStarts[object_nr]];
-  const u8* objectdata_end = &fifo_frame.fifoData[frame_info.objectEnds[object_nr]];
-  const u8* objectdata = objectdata_start;
-  const std::ptrdiff_t obj_offset =
-      objectdata_start - &fifo_frame.fifoData[frame_info.objectStarts[0]];
-
-  int cmd = *objectdata++;
-  int stream_size = Common::swap16(objectdata);
-  objectdata += 2;
-  QString new_label = QStringLiteral("%1:  %2 %3  ")
-                          .arg(obj_offset, 8, 16, QLatin1Char('0'))
-                          .arg(cmd, 2, 16, QLatin1Char('0'))
-                          .arg(stream_size, 4, 16, QLatin1Char('0'));
-  if (stream_size && ((objectdata_end - objectdata) % stream_size))
-    new_label += tr("NOTE: Stream size doesn't match actual data length\n");
-
-  while (objectdata < objectdata_end)
-    new_label += QStringLiteral("%1").arg(*objectdata++, 2, 16, QLatin1Char('0'));
-
-  m_detail_list->addItem(new_label);
-  m_object_data_offsets.push_back(0);
-
-  // Between objectdata_end and next_objdata_start, there are register setting commands
-  if (object_nr + 1 < static_cast<int>(frame_info.objectStarts.size()))
+  // Note that frame_info.objectStarts[object_nr] is the start of the primitive data,
+  // but we want to start with the register updates which happen before that.
+  const u32 object_start = (object_nr == 0 ? 0 : frame_info.objectEnds[object_nr - 1]);
+  const u32 object_size = frame_info.objectEnds[object_nr] - object_start;
```

Comment:
Ok... but now you are missing the register updates after the final object in the frame.

Also, the terminology is getting mixed up here. In the fifo playback code, Objects are explictly primitive draws and just the primitive draws. They are only defined so that the object skipping code can skip the objects and not the register writes around it.

Register writes being connected to objects is something that only happens in this UI code